File: src/tensorflow/ops/audio.nim
import ../core

{.compile:"audio/audio.cc".}


type DecodeWav* {.header:"/Users/martin//.nimble/pkgs/tensorflow-0.1.0/tensorflow/ops/audio/audio.h", importcpp:"DecodeWav/*'0*/".} = object
  operation*: Operation[ofloat]
  output*: ofloat

proc iidecodeWav(scope: Scope,
                contents: ostring,
                desired_channels: int64,
                desired_samples: int64): DecodeWav {.header:"/Users/martin//.nimble/pkgs/tensorflow-0.1.0/tensorflow/ops/audio/audio.h", importcpp:"DecodeWav(*#, #, #, #)", constructor.}

proc decodeWav*(scope: Scope,
                contents: ostring,
                desired_channels: int64 = -1.int,
                desired_samples: int64 = -1.int): DecodeWav =
  return iidecodeWav(scope,
                     contents,
                     desired_channels,
                     desired_samples)

converter decodeWavToOut*(op: DecodeWav): ofloat {.inline.} = return op.output


type EncodeWav* {.header:"/Users/martin//.nimble/pkgs/tensorflow-0.1.0/tensorflow/ops/audio/audio.h", importcpp:"EncodeWav/*'0*/".} = object
  operation*: Operation[ostring]
  output*: ostring

proc iiencodeWav(scope: Scope,
                audio: ofloat,
                sample_rate: oint32): EncodeWav {.header:"/Users/martin//.nimble/pkgs/tensorflow-0.1.0/tensorflow/ops/audio/audio.h", importcpp:"EncodeWav(*#, #, #)", constructor.}

proc encodeWav*(scope: Scope,
                audio: ofloat,
                sample_rate: oint32): EncodeWav =
  return iiencodeWav(scope,
                     audio,
                     sample_rate)

converter encodeWavToOut*(op: EncodeWav): ostring {.inline.} = return op.output

File: src/tensorflow/ops/audio/audio.cc
#include "tensorflow/cc/ops/const_op.h"
#include "../../ops/audio/audio.h"

DecodeWav::DecodeWav(tensorflow::Scope& scope, 
           tensorflow::Input contents, 
           int64_t desired_channels, 
           int64_t desired_samples) {
      if (!scope.ok())
          return;
      auto _contents = ::tensorflow::ops::AsNodeOut(scope, contents);
      ::tensorflow::Node *ret;
      const auto unique_name = scope.GetUniqueNameForOp("DecodeWav");
      auto builder = ::tensorflow::NodeBuilder(unique_name, "DecodeWav")
                                   .Input(_contents)
                                   .Attr("desired_channels", desired_channels)
                                   .Attr("desired_samples", desired_samples)
      ;
      scope.UpdateBuilder(&builder);
      scope.UpdateStatus(builder.Finalize(scope.graph(), &ret));
      if (!scope.ok()) return;
      scope.UpdateStatus(scope.DoShapeInference(ret));
      this->operation = ::tensorflow::Operation(ret);
      this->output = tensorflow::Output(ret, 0);
}

EncodeWav::EncodeWav(tensorflow::Scope& scope, 
           tensorflow::Input audio, 
           tensorflow::Input sample_rate) {
      if (!scope.ok())
          return;
      auto _audio = ::tensorflow::ops::AsNodeOut(scope, audio);
      if (!scope.ok())
          return;
      auto _sample_rate = ::tensorflow::ops::AsNodeOut(scope, sample_rate);
      ::tensorflow::Node *ret;
      const auto unique_name = scope.GetUniqueNameForOp("EncodeWav");
      auto builder = ::tensorflow::NodeBuilder(unique_name, "EncodeWav")
                                   .Input(_audio)
                                   .Input(_sample_rate)
      ;
      scope.UpdateBuilder(&builder);
      scope.UpdateStatus(builder.Finalize(scope.graph(), &ret));
      if (!scope.ok()) return;
      scope.UpdateStatus(scope.DoShapeInference(ret));
      this->operation = ::tensorflow::Operation(ret);
      this->output = tensorflow::Output(ret, 0);
}

